File: Vienna Tests/models/ArticleTests.m
//
//  ArticleTests.m
//  Vienna
//
//  Copyright © 2016 uk.co.opencommunity. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "Article.h"

static NSString * const GUID = @"07f446d2-8d6b-4d99-b488-cebc9eac7c33";
static NSString * const Author = @"Author McAuthorface";
static NSString * const Title = @"Lorem ipsum dolor sit amet";
static NSString * const Link = @"http://www.vienna-rss.org";
static NSString * const Enclosure = @"http://vienna-rss.sourceforge.net/img/vienna_logo.png";
static NSString * const EnclosureFilename = @"vienna_logo.png"; // last path component of Enclosure
static NSString * const Body =
    @"<p><strong>Pellentesque habitant morbi tristique</strong> senectus et netus "
    "et malesuada fames ac turpis egestas. Vestibulum tortor quam, feugiat vitae, "
    "ultricies eget, tempor sit amet, ante. Donec eu libero sit amet quam egestas semper."
    "<em>Aenean ultricies mi vitae est.</em> Mauris placerat eleifend leo. Quisque sit amet "
    "est et sapien ullamcorper pharetra. Vestibulum erat wisi, condimentum sed, <code>commodo "
    "vitae</code>, ornare sit amet, wisi. Aenean fermentum, elit eget tincidunt condimentum, "
    "eros ipsum rutrum orci, sagittis tempus lacus enim ac dui. "
    "<a href=\"#\">Donec non enim</a> in turpis pulvinar facilisis. Ut felis.</p>";


@interface ArticleTests : XCTestCase

@property (nonatomic, strong) Article *article;

@end

@implementation ArticleTests

- (void)setUp
{
    [super setUp];

    self.article = [[Article alloc] initWithGuid:GUID];
}

- (void)testAccessInstanceVariablesDirectly
{
    XCTAssertFalse([Article accessInstanceVariablesDirectly]);
}

#pragma mark - Test custom setters

- (void)testTitle
{
    self.article.title = Title;

    XCTAssertEqualObjects(self.article.title, Title);
}

- (void)testAuthor
{
    self.article.author = Author;

    XCTAssertEqualObjects(self.article.author, Author);
}

- (void)testLink
{
    self.article.link = Link;

    XCTAssertEqualObjects(self.article.link, Link);
}

- (void)testDate
{
    NSDate *date = [NSDate date];

    self.article.date = date;

    XCTAssertEqualObjects(self.article.date, date);
}

- (void)testDateCreated
{
    NSDate *date = [NSDate date];

    self.article.createdDate = date;

    XCTAssertEqualObjects(self.article.createdDate, date);
}

- (void)testBody
{
    self.article.body = Body;

    XCTAssertEqualObjects(self.article.body, Body);
}

- (void)testEnclosure
{
    self.article.enclosure = Enclosure;

    XCTAssertEqualObjects(self.article.enclosure, Enclosure);
}

- (void)testEnclosureRemoval
{
    self.article.enclosure = nil;

    XCTAssertNil(self.article.enclosure);
}

- (void)testHasEnclosure
{
    self.article.hasEnclosure = YES;

    XCTAssert(self.article.hasEnclosure);
}

- (void)testFolderId
{
    NSInteger folderId = 111;

    self.article.folderId = folderId;

    XCTAssertEqual(self.article.folderId, folderId);
}

- (void)testGuid
{
    self.article.guid = GUID;

    XCTAssertEqualObjects(self.article.guid, GUID);
}

- (void)testParentId
{
    NSInteger parentId = 222;

    self.article.parentId = parentId;

    XCTAssertEqual(self.article.parentId, parentId);
}

- (void)testStatus
{
    NSInteger status = ArticleStatusNew;

    self.article.status = status;

    XCTAssertEqual(self.article.status, status);
}

- (void)testMarkRead
{
    XCTAssertFalse(self.article.isRead);

    [self.article markRead:YES];

    XCTAssert(self.article.isRead);
}

- (void)testMarkRevised
{
    XCTAssertFalse(self.article.isRevised);

    [self.article markRevised:YES];

    XCTAssert(self.article.isRevised);
}

- (void)testMarkDeleted
{
    XCTAssertFalse(self.article.isDeleted);

    [self.article markDeleted:YES];

    XCTAssert(self.article.isDeleted);
}

- (void)testMarkFlagged
{
    XCTAssertFalse(self.article.isFlagged);

    [self.article markFlagged:YES];

    XCTAssert(self.article.isFlagged);
}

- (void)testMarkEnclosureDowloaded
{
    XCTAssertFalse(self.article.enclosureDownloaded);

    [self.article markEnclosureDownloaded:YES];

    XCTAssert(self.article.enclosureDownloaded);
}

- (void)testCompatibilityDate
{
    NSDate *date = [NSDate date];
    NSString *dateKeyPath = [@"articleData." stringByAppendingString:MA_Field_Date];

    self.article.date = date;

    XCTAssertEqualObjects([self.article valueForKeyPath:dateKeyPath], date);
}

- (void)testCompatibilityAuthor
{
    NSString *authorKeyPath = [@"articleData." stringByAppendingString:MA_Field_Author];

    self.article.author = Author;

    XCTAssertEqualObjects([self.article valueForKeyPath:authorKeyPath], Author);
}

- (void)testCompatibilitySubject
{
    NSString *subject = @"Lorem ipsum dolor sit amet";
    NSString *subjectKeyPath = [@"articleData." stringByAppendingString:MA_Field_Subject];

    self.article.title = subject;

    XCTAssertEqualObjects([self.article valueForKeyPath:subjectKeyPath], subject);
}

- (void)testCompatibilityLink
{
    NSString *link = @"http://www.vienna-rss.org";
    NSString *linkKeyPath = [@"articleData." stringByAppendingString:MA_Field_Link];

    self.article.link = link;

    XCTAssertEqualObjects([self.article valueForKeyPath:linkKeyPath], link);
}

- (void)testCompatibilitySummary
{
    NSString *summary = @"Lorem ipsum dolor sit amet";
    NSString *summaryKeyPath = [@"articleData." stringByAppendingString:MA_Field_Summary];

    self.article.body = summary;

    XCTAssertEqualObjects([self.article valueForKeyPath:summaryKeyPath], summary);
}

- (void)testRandomCompatibilityKeyPath
{
    NSString *randomArticleDataKeyPath = [@"articleData." stringByAppendingString:@"dummyProperty"];

    XCTAssertThrowsSpecificNamed([self.article valueForKeyPath:randomArticleDataKeyPath],
                                 NSException,
                                 NSUndefinedKeyException);
}

- (void)testRandomKeyPath
{
    NSString *randomKeyPath = @"dummyProperty";

    XCTAssertThrowsSpecificNamed([self.article valueForKeyPath:randomKeyPath],
                                 NSException,
                                 NSUndefinedKeyException);
}

- (void)testDescription
{
    NSString *title = @"Lorem ipsum dolor sit amet";

    self.article.guid = GUID;
    self.article.title = title;

    NSString *expectedDescription =
        [NSString stringWithFormat:@"{GUID=%@ title=\"%@\"", GUID, title];

    XCTAssertEqualObjects(self.article.description, expectedDescription);
}

#pragma mark - Expand tags

- (void)testExpandLinkTag
{
    NSString *string = @"$ArticleLink$";

    NSString *expectedString = [NSString stringWithFormat:@"%@/", Link];

    self.article.link = Link;

    NSString *expandedString = [self.article expandTags:string withConditional:YES];

    XCTAssertEqualObjects(expandedString, expectedString);
}

- (void)testExpandTitleTag
{
    NSString *string = @"$ArticleTitle$";
    NSString *expectedString = Title;

    self.article.title = Title;

    NSString *expandedString = [self.article expandTags:string withConditional:YES];

    XCTAssertEqualObjects(expandedString, expectedString);
}

- (void)testExpandArticleBodyTag
{
    NSString *string = @"$ArticleBody$";
    NSString *expectedString = Body;

    self.article.body = Body;

    NSString *expandedString = [self.article expandTags:string withConditional:YES];

    XCTAssertEqualObjects(expandedString, expectedString);
}

- (void)testExpandArticleAuthorTag
{
    NSString *string = @"$ArticleAuthor$";
    NSString *expectedString = Author;

    self.article.author = Author;

    NSString *expandedString = [self.article expandTags:string withConditional:YES];

    XCTAssertEqualObjects(expandedString, expectedString);
}

- (void)testExpandArticleEnclosureLinkTag
{
    NSString *string = @"$ArticleEnclosureLink$";
    NSString *expectedString = Enclosure;

    self.article.enclosure = Enclosure;

    NSString *expandedString = [self.article expandTags:string withConditional:YES];

    XCTAssertEqualObjects(expandedString, expectedString);
}

- (void)testExpandArticleEnclosureFileName
{
    NSString *string = @"$ArticleEnclosureFilename$";
    NSString *expectedString = EnclosureFilename;

    self.article.enclosure = Enclosure;

    NSString *expandedString = [self.article expandTags:string withConditional:YES];

    XCTAssertEqualObjects(expandedString, expectedString);
}

@end

File: Vienna Tests/ArticleTests.swift
//
//  ArticleTests.swift
//  Vienna Tests
//
//  Copyright 2020
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  https://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import XCTest

class ArticleTests: XCTestCase {

    let guid = "07f446d2-8d6b-4d99-b488-cebc9eac7c33"
    let author = "Author McAuthorface"
    let title = "Lorem ipsum dolor sit amet"
    let link = "http://www.vienna-rss.com"
    let enclosure = "http://vienna-rss.sourceforge.net/img/vienna_logo.png"
    let enclosureFilename = "vienna_logo.png" // last path component of Enclosure
    let body = """
    <p><strong>Pellentesque habitant morbi tristique</strong> senectus et netus
    et malesuada fames ac turpis egestas. Vestibulum tortor quam, feugiat vitae,
    ultricies eget, tempor sit amet, ante. Donec eu libero sit amet quam egestas
    semper. <em>Aenean ultricies mi vitae est.</em> Mauris placerat eleifend
    leo. Quisque sit amet est et sapien ullamcorper pharetra. Vestibulum erat
    wisi, condimentum sed, <code>commodo vitae</code>, ornare sit amet, wisi.
    Aenean fermentum, elit eget tincidunt condimentum, eros ipsum rutrum orci,
    sagittis tempus lacus enim ac dui. <a href="#">Donec non enim</a> in turpis
    pulvinar facilisis. Ut felis.</p>
    """

    var article: Article!
    var articleConverter: WebKitArticleConverter!

    override func setUpWithError() throws {
        try super.setUpWithError()
        // Put setup code here. This method is called before the invocation of each test method in the class.
        self.article = Article(guid: guid)
        self.articleConverter = WebKitArticleConverter()
    }

    override func tearDownWithError() throws {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        self.article = nil
        try super.tearDownWithError()
    }

    // MARK: Article Tests

    func testAccessInstanceVariablesDirectly() {
        XCTAssertFalse(Article.accessInstanceVariablesDirectly)
    }

    // MARK: - Test custom setters

    func testTitle() {
        self.article.title = title

        XCTAssertEqual(self.article.title, title)
    }

    func testAuthor() {
        self.article.author = author

        XCTAssertEqual(self.article.author, author)
    }

    func testLink() {
        self.article.link = link

        XCTAssertEqual(self.article.link, link)
    }

    func testDate() {
        let date = Date()

        self.article.date = date

        XCTAssertEqual(self.article.date, date)
    }

    func testDateCreated() {
        let date = Date()

        self.article.createdDate = date

        XCTAssertEqual(self.article.createdDate, date)
    }

    func testBody() {
        self.article.body = body

        XCTAssertEqual(self.article.body, body)
    }

    func testEnclosure() {
        self.article.enclosure = enclosure

        XCTAssertEqual(self.article.enclosure, enclosure)
    }

    func testEnclosureRemoval() {
        self.article.enclosure = nil

        XCTAssertNil(self.article.enclosure)
    }

    func testHasEnclosure() {
        self.article.hasEnclosure = true

        XCTAssert(self.article.hasEnclosure)
    }

    func testFolderId() {
        let folderId = 111

        self.article.folderId = folderId

        XCTAssertEqual(self.article.folderId, folderId)
    }

    func testGuid() {
        self.article.guid = guid

        XCTAssertEqual(self.article.guid, guid)
    }

    func testParentId() {
        let parentId = 222

        self.article.parentId = parentId

        XCTAssertEqual(self.article.parentId, parentId)
    }

    func testStatus() {
        let status = ArticleStatus.new.rawValue

        self.article.status = status

        XCTAssertEqual(self.article.status, status)
    }

    func testMarkRead() {
        XCTAssertFalse(self.article.isRead)

        self.article.markRead(true)

        XCTAssert(self.article.isRead)
    }

    func testMarkRevised() {
        XCTAssertFalse(self.article.isRevised)

        self.article.markRevised(true)

        XCTAssert(self.article.isRevised)
    }

    func testMarkDeleted() {
        XCTAssertFalse(self.article.isDeleted)

        self.article.markDeleted(true)

        XCTAssert(self.article.isDeleted)
    }

    func testMarkFlagged() {
        XCTAssertFalse(self.article.isFlagged)

        self.article.markFlagged(true)

        XCTAssert(self.article.isFlagged)
    }

    func testMarkEnclosureDowloaded() {
        XCTAssertFalse(self.article.enclosureDownloaded)

        self.article.markEnclosureDownloaded(true)

        XCTAssert(self.article.enclosureDownloaded)
    }

    func testCompatibilityDate() {
        let date = Date()
        let dateKeyPath = "articleData." + MA_Field_Date

        self.article.date = date

        XCTAssertEqual(self.article.value(forKeyPath: dateKeyPath) as? Date, date)
    }

    func testCompatibilityAuthor() {
        let authorKeyPath = "articleData." + MA_Field_Author

        self.article.author = author

        XCTAssertEqual(self.article.value(forKeyPath: authorKeyPath) as? String, author)
    }

    func testCompatibilitySubject() {
        let subject = "Lorem ipsum dolor sit amet"
        let subjectKeyPath = "articleData." + MA_Field_Subject

        self.article.title = subject

        XCTAssertEqual(self.article.value(forKeyPath: subjectKeyPath) as? String, subject)
    }

    func testCompatibilityLink() {
        let link = "http://www.vienna-rss.com"
        let linkKeyPath = "articleData." + MA_Field_Link

        self.article.link = link

        XCTAssertEqual(self.article.value(forKeyPath: linkKeyPath) as? String, link)
    }

    func testCompatibilitySummary() {
        let summary = "Lorem ipsum dolor sit amet"
        let summaryKeyPath = "articleData." + MA_Field_Summary

        self.article.body = summary

        XCTAssertEqual(self.article.value(forKeyPath: summaryKeyPath) as? String, summary)
    }

//    func testRandomCompatibilityKeyPath() {
//        let randomArticleDataKeyPath = "articleData.dummyProperty"
//
//        XCTAssertThrowsSpecificNamed(self.article.value(forKeyPath: randomArticleDataKeyPath), NSException, NSUndefinedKeyException);
//    }

//    func testRandomKeyPath() {
//        let randomKeyPath = "dummyProperty"
//
//        XCTAssertThrowsSpecificNamed(self.article.value(forKeyPath: randomKeyPath), NSException, NSUndefinedKeyException);
//    }

    func testDescription() {
        let title = "Lorem ipsum dolor sit amet"

        self.article.guid = guid
        self.article.title = title

        let expectedDescription = String(format: "{GUID=%@ title=\"%@\"", guid, title)

        XCTAssertEqual(self.article.description, expectedDescription)
    }

    // MARK: - Expand tags

    func testExpandLinkTag() {
        let string = "$ArticleLink$/development"

        let expectedString = link + "/development"

        self.article.link = link

        let expandedString = self.articleConverter.expandTags(of: self.article, intoTemplate: string, withConditional: true)

        XCTAssertEqual(expandedString, expectedString)
    }

    func testExpandTitleTag() {
        let string = "$ArticleTitle$"
        let expectedString = title

        self.article.title = title

        let expandedString = self.articleConverter.expandTags(of: self.article, intoTemplate: string, withConditional: true)
        XCTAssertEqual(expandedString, expectedString)
    }

    func testExpandArticleBodyTag() {
        let string = "$ArticleBody$"
        let expectedString = body

        self.article.body = body

        let expandedString = self.articleConverter.expandTags(of: self.article, intoTemplate: string, withConditional: true)

        XCTAssertEqual(expandedString, expectedString)
    }

    func testExpandArticleAuthorTag() {
        let string = "$ArticleAuthor$"
        let expectedString = author

        self.article.author = author

        let expandedString = self.articleConverter.expandTags(of: self.article, intoTemplate: string, withConditional: true)

        XCTAssertEqual(expandedString, expectedString)
    }

    func testExpandArticleEnclosureLinkTag() {
        let string = "$ArticleEnclosureLink$"
        let expectedString = enclosure

        self.article.enclosure = enclosure

        let expandedString = self.articleConverter.expandTags(of: self.article, intoTemplate: string, withConditional: true)

        XCTAssertEqual(expandedString, expectedString)
    }

    func testExpandArticleEnclosureFileName() {
        let string = "$ArticleEnclosureFilename$"
        let expectedString = enclosureFilename

        self.article.enclosure = enclosure

        let expandedString = self.articleConverter.expandTags(of: self.article, intoTemplate: string, withConditional: true)

        XCTAssertEqual(expandedString, expectedString)
    }

    func testURLIDNinArticleView() {

        self.article.link = "http://ουτοπία.δπθ.gr/نجيب_محفوظ/"
        let htmlTextFromIDNALink = self.articleConverter.articleText(from: [ self.article ])

        self.article.link = "http://xn--kxae4bafwg.xn--pxaix.gr/%D9%86%D8%AC%D9%8A%D8%A8_%D9%85%D8%AD%D9%81%D9%88%D8%B8/"
        let htmlTextFromResolvedIDNALink = self.articleConverter.articleText(from: [ self.article ])

        XCTAssertEqual(htmlTextFromIDNALink, htmlTextFromResolvedIDNALink)
    }

}
